jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java

```diff
@@ -0,0 +1,432 @@
+     * <p>The set of returned priorities should be stable over
+     * time, typically constrained in the range {@code 0-10}.</p>
+     *
+     * @param request the suspended request to compute the priority for
+     * @return the priority of the given suspended request, a value {@code >= 0}
+     */
+    protected int getPriority(Request request)
+    {
+        return 0;
+    }
+
+    /**
+     * <p>Fails the given suspended request/response with the given error code and failure.</p>
+     * <p>This method is called only for suspended requests, in case of timeout while suspended,
+     * or in case of failure when trying to handle a resumed request.</p>
+     *
+     * @param request the request to fail
+     * @param response the response to fail
+     * @param callback the callback to complete
+     * @param status the failure status code
+     * @param failure the failure
+     */
+    protected void failSuspended(Request request, Response response, Callback callback, int status, Throwable failure)
+    {
+        Response.writeError(request, response, callback, status, null, failure);
+    }
+
+    private boolean forward(Request request, Response response, Callback callback) throws Exception
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} forwarding {}", this, request);
+        request.addHttpStreamWrapper(stream -> new Resumer(stream, request));
+        return super.handle(request, response, callback);
+    }
+
+    private void suspend(Request request, Response response, Callback callback)
+    {
+        int priority = Math.max(0, getPriority(request));
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} suspending priority={} {}", this, priority, request);
+        Entry entry = new Entry(request, response, callback, priority);
+        queues.compute(priority, (k, v) ->
+        {
+            if (v != null)
+                return v;
+            priorities.add(priority);
+            return new ConcurrentLinkedQueue<>();
+        }).offer(entry);
+        timeouts.schedule(entry);
+    }
+
+    private void resume()
+    {
+        // See correspondent state machine logic in handle() and expire().
+        int permits = state.getAndIncrement();
+        if (permits >= 0)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} no suspended requests to resume", this);
+            return;
+        }
+
+        while (true)
+        {
+            if (resumeSuspended())
+                return;
+
+            // Found no suspended requests yet, but there will be.
+            // This covers the small race window in handle(), where
+            // the state is updated and then the request suspended.
+            Thread.onSpinWait();
+        }
+    }
+
+    private boolean resumeSuspended()
+    {
+        for (Integer priority : priorities)
+        {
+            Queue<Entry> queue = queues.get(priority);
+            Entry entry = queue.poll();
```

Comment:
There is a tiny window for a NPE here as a new priority is added to the `priorities` just before it creates the new entry in `queues`, thus `queue` might end up being null here.
If it is, I think it is safe to return false, as the caller will just spin and call again

Reply:
@gregw we also might see the queue but not the entry, so I modified the `offer()` side too.